### PR TITLE
bugfix(behavior): Fix unexpected propaganda influence from contained units with direct Propaganda Tower behavior modules

### PR DESCRIPTION
**Pull request by Stubbjax**
This change fixes an issue where propaganda effects persist for contained units that have a direct `PropagandaTowerBehavior` module (e.g. Attack Outpost, Assault Troop Crawler).

The contained tower logic was only ever anticipated to be used for the Overlord's tower attachment, which is technically considered as a 'contained' passenger by the Overlord - hence the need for the two container checks. However, as the `PropagandaTowerBehavior` module can be added directly to units themselves, the container check does not function correctly for singular containers.

In the 1.04 example below, an Assault Troop Crawler enters a tunnel and its propaganda effect persists for units around the tunnel:

image

### CODE REVIEW

File: Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp [hunk 1]
```diff
@@ -205,7 +205,15 @@ UpdateSleepTime PropagandaTowerBehavior::update( void )
 		}
 	}
 
-	if( self->getContainedBy()  &&  self->getContainedBy()->getContainedBy() )
+#if RETAIL_COMPATIBLE_CRC
+	Bool contained = self->getContainedBy() && self->getContainedBy()->getContainedBy();
+#else
+	Bool contained = (self->isKindOf(KINDOF_PORTABLE_STRUCTURE)) ?
+		contained = self->getContainedBy() && self->getContainedBy()->getContainedBy() :
+		contained = self->getContainedBy();
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: Tertiary operator is written like so

```cpp
Bool contained = (self->isKindOf(KINDOF_PORTABLE_STRUCTURE)) ?
		self->getContainedBy() && self->getContainedBy()->getContainedBy() :
		self->getContainedBy();
```

File: Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp [hunk 2]
```diff
@@ -205,7 +205,15 @@ UpdateSleepTime PropagandaTowerBehavior::update( void )
 		}
 	}
 
-	if( self->getContainedBy()  &&  self->getContainedBy()->getContainedBy() )
+#if RETAIL_COMPATIBLE_CRC
+	Bool contained = self->getContainedBy() && self->getContainedBy()->getContainedBy();
+#else
+	Bool contained = (self->isKindOf(KINDOF_PORTABLE_STRUCTURE)) ?
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: Can we implement this more generic, without kindof test? I expect in its current form there could be edge cases where this check fails.

Comment (Stubbjax) [reply Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: This change does cover all cases in its current form, given that objects of type `PORTABLE_STRUCTURE` are always considered contained.

There may be an overloaded function somewhere that portable structures have to distinguish them from other passengers that could be used here. Will need further investigation.

**Comment by xezon** [comment]
This change correctly identifies the cause of the issue, but the fix is not properly implemented.

The KindOf inside the Contain module is configurable in INI:

```
  Behavior = HelixContain ModuleTag_09
    Slots                     = 1
    DamagePercentToUnits      = 100%
    AllowInsideKindOf         = PORTABLE_STRUCTURE
    PassengersAllowedToFire   = Yes
  End
```

Therefore, hardcoding this KindOf is not right.

The correct way to test is in the contain module via `ContainModuleInterface::isSpecialOverlordStyleContainer`. An example of how this is used can be seen in `TransportContain::isPassengerAllowedToFire`.

Therefore, I have created new change #1609.

**Comment by Stubbjax** [comment]
Oh right, `isSpecialOverlordStyleContainer` is what I was looking for, though it hardly feels any more generic.

**Comment by Stubbjax** [comment]
I believe I have figured out a more comprehensive solution that covers all cases.

https://github.com/user-attachments/assets/2ea13a3c-0257-4a1f-807b-785a2fb52f41

File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp [hunk 3]
```diff
@@ -207,7 +208,18 @@ UpdateSleepTime PropagandaTowerBehavior::update( void )
...
 	}
 
-	if( self->getContainedBy()  &&  self->getContainedBy()->getContainedBy() )
+#if RETAIL_COMPATIBLE_CRC
+	Bool contained = self->getContainedBy() && self->getContainedBy()->getContainedBy();
+#else
+	Object* container = self->getContainedBy();
+
+	while (container && container->getContainedBy())
+		container = container->getContainedBy();
```
Comment (xezon) [inline GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: I have seen this pattern a few times now and I think it warrants adding a new function in class Object: `Object* getTopContainedBy()`

File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp [hunk 4]
```diff
@@ -207,7 +208,18 @@ UpdateSleepTime PropagandaTowerBehavior::update( void )
...
+
+	Bool contained = (container && container->getContain()->isEnclosingContainerFor(self));
+#endif
+
+	if (contained)
 	{
 		// If our container is contained, we turn the heck off.  Seems like a weird specific check, but all of
```
Comment (xezon) [inline GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: Please update/append to this comment here. The situation is more complicated now.

File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp [hunk 5]
```diff
@@ -207,7 +208,18 @@ UpdateSleepTime PropagandaTowerBehavior::update( void )
...
-	if( self->getContainedBy()  &&  self->getContainedBy()->getContainedBy() )
+#if RETAIL_COMPATIBLE_CRC
+	Bool contained = self->getContainedBy() && self->getContainedBy()->getContainedBy();
+#else
+	Object* container = self->getContainedBy();
+
+	while (container && container->getContainedBy())
+		container = container->getContainedBy();
+
+	Bool contained = (container && container->getContain()->isEnclosingContainerFor(self));
```
Comment (xezon) [inline GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: I think this would not account for a Top container that is not enclosing and the child container is enclosing with a propaganda unit inside that. Is that possible?

Perhaps have a `getTopEnclosingContainedBy()` function.

Comment (Stubbjax) [reply GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: Ah yeah, I had thought about that. I believe it may just need to also check for the enclosing container within the loop condition and end when it is no longer enclosing.

File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp (hunk 3 above)
Comment (Stubbjax) [reply GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: I've simplified the logic somewhat so this might no longer be necessary.

File: Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp [hunk 6]
```diff
@@ -205,11 +206,28 @@ UpdateSleepTime PropagandaTowerBehavior::update( void )
...
 	}
 
-	if( self->getContainedBy()  &&  self->getContainedBy()->getContainedBy() )
+#if RETAIL_COMPATIBLE_CRC
+	Bool contained = self->getContainedBy() && self->getContainedBy()->getContainedBy();
+#else
+	Object* container = self->getContainedBy();
+	Bool contained = false;
+
+	while (container)
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: Can condense code with

```cpp
for (Object* container = self->getContainedBy(); container; container = container->getContainedBy())
```

File: Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp [hunk 7]
```diff
@@ -205,11 +206,28 @@ UpdateSleepTime PropagandaTowerBehavior::update( void )
...
+	{
+		ContainModuleInterface* containModule = container->getContain();
+		if (containModule && containModule->isEnclosingContainerFor(self))
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: This seems odd in the loop because `isEnclosingContainerFor(self)` would only ever be true for the direct parent container, no?

Comment (Stubbjax) [reply Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: Not necessarily; it returns whether the container _would_ enclose the passed object. But I think you're right, it makes sense to always pass the direct child.

File: Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp [hunk 8]
```diff
@@ -205,11 +206,25 @@ UpdateSleepTime PropagandaTowerBehavior::update( void )
...
+		}
+	}
+#endif
+
+	if (contained)
 	{
-		// If our container is contained, we turn the heck off.  Seems like a weird specific check, but all of
-		// attacking is guarded by the same check in isPassengersAllowedToFire.  We similarly work in a container,
-		// but not in a double container.
+		// If our container or any parent containers are an enclosing container, we turn the heck off.
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: I suggest

`// TheSuperHackers @bugfix If our container or any parent containers are enclosing, we turn the heck off.`

and move comment below `#else` ?

File: Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp [hunk 9]
```diff
@@ -205,11 +206,25 @@ UpdateSleepTime PropagandaTowerBehavior::update( void )
...
 
-	if( self->getContainedBy()  &&  self->getContainedBy()->getContainedBy() )
+#if RETAIL_COMPATIBLE_CRC
+	Bool contained = self->getContainedBy() && self->getContainedBy()->getContainedBy();
+#else
+	Bool contained = false;
+
+	for (Object* child = self, *container = self->getContainedBy(); container; child = container, container = container->getContainedBy())
+	{
+		ContainModuleInterface* containModule = container->getContain();
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: How about we simplify this code here by creating a `Object::getEnclosingContain()`? which only returns `ContainModuleInterface*` if it is enclosing?

Comment (xezon) [reply Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: I see you moved the whole loop into new function. What I meant was:

```cpp
ContainModuleInterface* Object::getEnclosingContain(Object* containee)
{
  ContainModuleInterface* containModule = getContain();
  if (containModule && containModule->isEnclosingContainerFor(containee))
  {
    return containModule;
  }
  return NULL;
}
```
This way it is more in line with what `getContain` is doing.

And then can loop like so:

```cpp
Object* Object::getFirstEnclosingContainedBy()
{
  for (Object* child = this, Object *container = getContainedBy(); container; child = container, container = container->getContainedBy())
  {
    if (container->getEnclosingContain(child))
      return container;
  }
  return NULL;
}
```

Basically keep
`getEnclosingContain` in line with `getContain`
`getFirstEnclosingContainedBy` in line with `getContainedBy`

I did not test it. Just draft.

Comment (Stubbjax) [reply Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: Is the current approach unsatisfactory?

> Basically keep
> `getEnclosingContain` in line with `getContain`
> `getFirstEnclosingContainedBy` in line with `getContainedBy`

Functionally, `getEnclosingContain` is more in line with `getContainedBy` as both functions look up the containment chain for an external container object, whereas `getContain` is an instance accessor akin to `getAI`, `getPhysics`, `getStealth`, etc.

Comment (xezon) [reply Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: It is a bit confusing because the original `getContain` returns the direct contain module of the object, whereas your proposed `getEnclosingContain` returns Null or the first parent contain module that is enclosing. Perhaps name this differently to more clearly conceptually distinguish from `getContain`?

Other than that, my proposed `getFirstEnclosingContainedBy` allows to get hands on the Object first, which could be useful for future uses, because `ContainModuleInterface` currently does not expose the owning object. There simply is much more we can do with an `Object` than just `ContainModuleInterface`, which makes this function much more versatile.

Comment (xezon) [reply Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: I think I now understand your thinking behind `getEnclosingContain`, from natural language we can assume that it is a parent contain. My monkey brain thinks more of a "get the enclosing contain module that is owned by this object".

Comment (xezon) [reply Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: I think my function would be clear with this name:
`ContainModuleInterface* Object::getContainIfEnclosing(Object* containee)`

Comment (Stubbjax) [reply Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: > It is a bit confusing because the original `getContain` returns the direct contain module of the object, whereas your proposed `getEnclosingContain` returns Null or the first parent contain module that is enclosing. Perhaps name this differently to more clearly conceptually distinguish from `getContain`?

It was meant to be comparable to `getContainedBy` rather than `getContain`, but I see how the naming can be misleading.

How about changing the return type to `Object*` and the name to `getEnclosingContainedBy` to better align with `getContainedBy`?

```diff
- ContainModuleInterface* getEnclosingContain();
+ Object* getEnclosingContainedBy();
```

Comment (xezon) [reply Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: Yes. I suggest implement it roughly the way I suggested above. Can optimize the names where applicable.

File: Generals/Code/GameEngine/Include/GameLogic/Object.h [hunk 10]
```diff
@@ -421,6 +421,7 @@ class Object : public Thing, public Snapshot
 	void onRemovedFrom( Object *removedFrom );
 	Int getTransportSlotCount() const;
 	void friend_setContainedBy( Object *containedBy ) { m_containedBy = containedBy; }
+	Object* getEnclosingContainedBy(); // <<< Find the first enclosing container in the containment chain.
```
Comment (xezon) [inline Generals/Code/GameEngine/Include/GameLogic/Object.h]: EA comment style is: `///<`

File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp [hunk 11]
```diff
@@ -207,11 +208,13 @@ UpdateSleepTime PropagandaTowerBehavior::update( void )
 		}
 	}
 
-	if( self->getContainedBy()  &&  self->getContainedBy()->getContainedBy() )
+#if RETAIL_COMPATIBLE_CRC
+	if (self->getContainedBy() && self->getContainedBy()->getContainedBy())
+#else
+	// If our container or any parent containers are enclosing, we turn the heck off.
+	if (self->getEnclosingContain())
```
Comment (xezon) [inline GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: I think this will not compile.

File: Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp [hunk 12]
```diff
@@ -205,11 +206,13 @@ UpdateSleepTime PropagandaTowerBehavior::update( void )
 		}
 	}
 
-	if( self->getContainedBy()  &&  self->getContainedBy()->getContainedBy() )
+#if RETAIL_COMPATIBLE_CRC
+	if (self->getContainedBy() && self->getContainedBy()->getContainedBy())
+#else
+	// If our container or any parent containers are enclosing, we turn the heck off.
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Behavior/PropagandaTowerBehavior.cpp]: `TheSuperHackers @bugfix`

File: GeneralsMD/Code/GameEngine/Include/GameLogic/Object.h [hunk 13]
```diff
@@ -446,6 +446,7 @@ class Object : public Thing, public Snapshot
 	void onRemovedFrom( Object *removedFrom );
 	Int getTransportSlotCount() const;
 	void friend_setContainedBy( Object *containedBy ) { m_containedBy = containedBy; }
+	Object* getEnclosingContainedBy(); // <<< Find the first enclosing container in the containment chain.
```
Comment (xezon) [inline GeneralsMD/Code/GameEngine/Include/GameLogic/Object.h]: ///<

Comment (Stubbjax) [reply GeneralsMD/Code/GameEngine/Include/GameLogic/Object.h]: Hmm, thought I checked both. Must have somehow looked at the same one twice.

**Review by xezon — APPROVED** [review]
